File: lc133_clone_graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    #return a list of current_node neighbors after cloning them, if they are not
    #cloned already, also updates the hash_table with these new clones. these neighbor_list
    #may/may_not have their neighbors set correctly
    def clone_neighbors(self, current_node, hash_table):
        n_list=[]
        for x in current_node.neighbors:
            if x not in hash_table:
                #clone
                n=Node(x.val, None)
                hash_table[x]=n
            #already cloned
            n_list+=[x]
        return n_list
                
        
    def cloneGraph(self, node: 'Node') -> 'Node':
        """
        A BFS solution: idea is to have a traverse_list, initially with the given node.
        Clone the node, update the traverse list with the neighbors of just cloned node.
        keep looping on traverse list, untill it becomes empty.
        while updating traverse_list ensure that duplicates dont go into it, hence a set is
        good idea.
        """
        
        if not node:
            #empty set
            return None
        elif not node.neighbors:
            return Node(node.val, None)
        
        hash_table=dict() #stores old and new graphs node mapping
        
        traverse_list=[(node, False)];
        current_node=traverse_list[0]
        while( True ):
            if current_node[0] in hash_table and not current_node[1]:
                #already cloned but not neighbored
                hash_table[current_node[0]].neighbors=self.clone_neighbors(current_node[0], hash_table)
                traverse_list.append((current_node, True))
            elif current_node[0] not in hash_table:#current_node not in hash_table
                t=self.clone_neighbors(current_node[0], hash_table)
                n=Node(current_node[0].val, t)
                hash_table[current_node[0]]= n
                traverse_list.append((current_node, True))
                for x in t:
                    traverse_list.append((t, False))
            else:
                #current node in hash_table and neighbored, check next
                try:
                    current_node=traverse_list.pop()
                except IndexError:
                    break
            
                
        return hash_table[node]

File: lc133_clone_graph.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     int numNeighbors;
 *     struct Node** neighbors;
 * };
 */


struct HashTable{
    struct Node* orig;
    struct Node* newg;
};
struct HashTable HashTable[100];//hashtable[0] will have first-node of graph

//returns node of the newGraph corresponding to the oldGraph, if present. otherwise NULL
struct Node* table_lookup(struct Node* node){
    int i=0;
    for(i=0; i< 100; i++){
        if (HashTable[i].orig==node)
            return HashTable[i].newg;
    }
    return NULL;
}
void init_table(){
    int i=0;
    for(i=0; i< 100; i++){
        HashTable[i].orig=NULL;
        HashTable[i].newg=NULL;
    }
}
void add_to_table(struct Node* orig, struct Node* newg){
    int i=0;
    for(i=0; i< 100; i++){
        if (HashTable[i].orig==NULL){
            HashTable[i].orig=orig;
            HashTable[i].newg=newg;
        }
    }
}

struct Node* makeNode(struct Node **orig){
    if (*orig == NULL) return NULL;
    struct Node* n ;
    if ((n=table_lookup(*orig)) != NULL){
        return n;
    }else{
        n= malloc(sizeof(struct Node));
        n->val =(*orig)->val;
        n->numNeighbors =(*orig)->numNeighbors;
        *(n->neighbors)=makeNode((*orig)->neighbors);
        add_to_table(*orig, n);
    }
    return n;
}
struct Iter{
    struct Node* node;
    struct Iter* next;
    int cloned;
};
void traverse_list_init(struct Iter **tRef, struct Node* s){
    struct Iter* tmp=malloc(sizeof(struct Iter));
            tmp->next=NULL;
            tmp->node=s;
            tmp->cloned=0;
            *tRef=tmp;
}
struct Node* traverse_list_pop(struct Iter **tRef){
    //returns the first node which is not yet cloned.
    struct Iter* t=*tRef;
    while(t){ 
        if(t->cloned){
            t=t->next;
        }else{
            t->cloned=1;
            return t->node;
        }     
    }
    return NULL;//NULL
}
void traverse_list_update(struct Iter **tRef, struct Node ** neighbor_nodes){
    //updates the iterator with neighbor_nodes
    struct Node* n= *neighbor_nodes;
    struct Iter* t=*tRef;
    //checks for duplicates
    int duplicate=0;
    while(n){
        duplicate=0;
        while(t){
            if(n->val==t->node->val){
                duplicate=1;
                break;
            } 
            t=t->next;
        }        
        if(duplicate == 0){//add to the head
            struct Iter* tmp=malloc(sizeof(struct Iter));
            tmp->next=*tRef;
            tmp->node=n;
            tmp->cloned=0;
            *tRef=tmp;
        }
        n=*(n->neighbors);
    }
}
void traverse_list_free(struct Iter **tRef){
    return;
}
struct Node *cloneGraph(struct Node *s) {
    
    if (s == NULL) return s;//empty graph
    
    /*idea is to have  a hash_table to keep track of cloned nodes.
    and maintain a traverse_list which keeps track of nodes traveresed so far.
    */
    init_table();//initailize the hashtable
    struct Iter* traverse_list=NULL;//to store nodes that needs to be traversed
    traverse_list_init(&traverse_list, s);//initially to first input graph node.
    struct Node* currentNode=NULL, *newGnode;
    while((currentNode = traverse_list_pop(&traverse_list)) != NULL){
        //currentNode->cloned=1;
        if ((newGnode=table_lookup(currentNode))==NULL){
            //hashtable doesnt have this node. clone this node and its neighbors
            //
            newGnode=makeNode(&currentNode);
            //add_to_table(currentNode, newGNode);
            //update traverse_list, so that other nodes can be processed on next iteration
            traverse_list_update(&traverse_list, currentNode->neighbors);//update traverse_list with all neighbors that are not in hashtable
            
        }//else hashtable has-this node so its already cloned.
    }
    traverse_list_free(&traverse_list);
    return table_lookup(s);
}
